Make incremental updates to CSS conventions documentation (#20)
hypothesis/frontend-toolkit: docs/css-style-guide.md

# Hypothesis CSS Style Guide

> To enable efficient and safe evolution of our front end projects, we follow a few basic guides for styling:

Author: Using the term "guides" here per discussion last week, as "principles" are slightly higher-level in that model.

Reviewer: "guides" or "guidelines"?

- Organize styles into components for encapsulation, ease of reuse and refactoring
- Extract reusable patterns into mixins whenever possible.
- Use simple, non-nested class selectors to avoid specificity problems.
- Be careful with CSS preprocessor features. Use mixins to avoid repetition and variables to ensure consistency. Avoid SASS' `@extends` feature.
- Use [naming](#naming) conventions to indicate which classes are purely presentational and which are referenced in code

> - Use modern CSS features (flexbox) for layout, preferably via layout-utility mixins

Author: This will apply once https://github.com/hypothesis/client/pull/2247 is merged.

> - Use mixins whenever possible within component SASS rules. If there is no applicable mixin,

Author: This is probably the most substantive change here...

- Only reference classes with `js-` and `is-` prefixes in JavaScript code
- Use names with a `js-` prefix for handle or "ref" classes that are used by JS to get a reference to particular DOM elements. These classes should not appear in stylesheets

## SASS Features
## Working in SASS

- Use variables, functions and mixins

> - [Avoid @extends](https://www.sitepoint.com/avoid-sass-extend/). Use mixins instead

Author: We said this already

Reviewer: I think it makes more sense to have the "avoid `@extends`" here rather than at the top though.

}
```

> ## Colors and Typography

Author: I condensed this section; let me know if that's controversial.

  rather than hard-coding CSS rule values in component CSS
- We use flexbox instead for layout; there are some utility mixins and classes
  to make component styling more convenient.
- Avoid vendor prefixes ([autoprefixer](https://github.com/postcss/autoprefixer) will
  add these automatically).

## Browser Support

TODO needs update

> - Hypothesis projects should be functional on Internet Explorer 11 and above. CSS features that are only available in newer browsers can still be used but the site should behave gracefully in older browsers

Reviewer: Can you update this section based on https://web.hypothes.is/help/which-browsers-are-supported-by-hypothesis/? It says:

> The Hypothesis Chrome extension, Bookmarklet, and LMS application work within any modern web browser (a version of Chrome, Firefox, Safari, Microsoft Edge, or equivalent released within the last 12 months). Note that as of 1 July 2020, Hypothesis will no longer support Internet Explorer.

}
```

- Use color, typography and other values defined in the shared `variables` module
  (or, even better: use a utility mixin) rather than hard-coding CSS rule values
  in component CSS
- We use flexbox instead for layout; there are some utility mixins and classes
  to make component styling more convenient.
- Avoid vendor prefixes ([autoprefixer](https://github.com/postcss/autoprefixer) will
  add these automatically).

## Browser Support

> Hypothesis applications should work within any modern web browser (a version of Chrome, Firefox, Safari, Microsoft Edge, or equivalent released within the last 12 months). Note that as of 1 July 2020, Hypothesis will no longer support Internet Explorer.

Reviewer: ```suggestion
Hypothesis applications should work within any modern web browser (a version of Chrome, Firefox, Safari, Microsoft Edge, or equivalent released within the last 12 months). Note that as of 1 July 2020, Hypothesis does not support Internet Explorer.
```

Since 1st July 2020 is coming up very soon, we can pretend it is in the past here.
